pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,178 @@
+}
+
+func prepDownscale(ctx context.Context, logger log.Logger, ar v1.AdmissionReview, api kubernetes.Interface, client httpClient) *v1.AdmissionResponse {
+	logger = log.With(logger, "name", ar.Request.Name, "resource", ar.Request.Resource.Resource, "namespace", ar.Request.Namespace)
+
+	oldObj, oldGVK, err := codecs.UniversalDeserializer().Decode(ar.Request.OldObject.Raw, nil, nil)
+	if err != nil {
+		return allowErr(logger, "can't decode old object, allowing the change", err)
+	}
+	logger = log.With(logger, "request_gvk", oldGVK)
+
+	oldReplicas, err := replicas(oldObj, oldGVK)
+	if err != nil {
+		return allowErr(logger, "can't get old replicas, allowing the change", err)
+	}
+	logger = log.With(logger, "old_replicas", int32PtrStr(oldReplicas))
+
+	newObj, newGVK, err := codecs.UniversalDeserializer().Decode(ar.Request.Object.Raw, nil, nil)
+	if err != nil {
+		return allowErr(logger, "can't decode new object, allowing the change", err)
+	}
+
+	newReplicas, err := replicas(newObj, newGVK)
+	if err != nil {
+		return allowErr(logger, "can't get new replicas, allowing the change", err)
+	}
+	logger = log.With(logger, "new_replicas", int32PtrStr(newReplicas))
+
+	// Both replicas are nil, nothing to warn about.
+	if oldReplicas == nil && newReplicas == nil {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	// Changes from/to nil scale are not downscales strictly speaking.
+	if oldReplicas == nil || newReplicas == nil {
+		return allowWarn(logger, "old/new replicas is nil, allowing the change")
+	}
+	// If it's not a downscale, just log debug.
+	if *oldReplicas < *newReplicas {
+		level.Debug(logger).Log("msg", "upscale allowed")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	if *oldReplicas == *newReplicas {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	// Get the resource labels: for example, for a StatefulSet, it will be the labels of the StatefulSet itself,
+	// while for a Scale object, it will be the labels of the StatefulSet/Deployment/ReplicaSet that the Scale object belongs to.
+	var lbls map[string]string
+	switch o := oldObj.(type) {
+	case *appsv1.Deployment:
+		lbls = o.Labels
+	case *appsv1.StatefulSet:
+		lbls = o.Labels
+	case *appsv1.ReplicaSet:
+		lbls = o.Labels
+	case *autoscalingv1.Scale:
+		lbls, err = getResourceLabels(ctx, ar, api)
+		if err != nil {
+			return allowBecauseCannotGetResource(ar, logger, err)
+		}
+	default:
+		return allowWarn(logger, fmt.Sprintf("unsupported type %T, allowing the change", o))
+	}
+
+	// Since it's a downscale, check if the resource has the label that indicates it's ready to be prepared to be downscaled.
+	// Create a slice of endpoint addresses for pods to send HTTP post requests to and to fail if any don't return 200
+	if !*ar.Request.DryRun && lbls[PrepDownscaleLabelKey] == PrepDownscaleLabelValue {
```

Comment:
Nit, but I'd do a shortcut here:

```go
	if lbls[PrepDownscaleLabelKey] != PrepDownscaleLabelValue {
		// Not labeled, nothing to do.
		return &v1.AdmissionResponse{Allowed: true}
	}
	if *ar.Request.DryRun {
		// TODO: check whether downscale is possible (was another zone downscaled less than allowed time before?)
		// Respond based on that.
	}	
```

Reply:
Good idea. This has been implemented as such.